### PR TITLE
Script to get v4 genome release HT

### PR DESCRIPTION
**Pull request by KoalaQin**
I copied from Mike's PR #477 (thanks to Mike's and KC's hardwork on this) and adapted it to work for v4 genomes release. 
There are a few points worth noting: 

1. freq_ht needs to drop all subsets except HGDP/TGP; 
2. info_ht was from v3, it has a different schema, there's no `allele_info` field, `info` contains most `site_info` except "SOR" and "AS_SOR", may need to find them somewhere. 
3. all the changes in other files except `gnomad_qc/v4/create_release/create_release_sites_ht_genomes.py` will be reverted once the exome release PR is in. 

Test run is here: [8d33837b811a458b9307db9e116fe7b2](https://console.cloud.google.com/dataproc/jobs/8d33837b811a458b9307db9e116fe7b2?region=us-central1&authuser=0&project=broad-mpg-gnomad)

### CODE REVIEW

File: gnomad_qc/v4/create_release/create_release_sites_ht_genomes.py [hunk 1]
```diff
@@ -0,0 +1,790 @@
...
+        freq_meta=freq_meta,
+        freq_index_dict=make_freq_index_dict_from_meta(hl.literal(freq_meta)),
+        freq_meta_sample_count=array_exprs["freq_meta_sample_count"],
+    )
+
+    return freq_ht
+
+
+# Config is added as a function, so it is not evaluated until the function is called.
+def get_config(
```
Comment (jkgoodrich) [inline gnomad_qc/v4/create_release/create_release_sites_ht_genomes.py:166]: So far from what I see this is exactly the same as Mike's and can easily just have a `data_type: str = "exomes",` parameter added where all the places that have `data_type="genomes"` can just be changed to `data_type=data_type` and then the genomes and exomes release can use the same config

File: gnomad_qc/v4/create_release/create_release_sites_ht_genomes.py [hunk 2]
```diff
@@ -0,0 +1,790 @@
...
+                "select": [r for r in release_sites(data_type="genomes").ht().row],
+                "select_globals": [
+                    g for g in release_sites(data_type="genomes").ht().globals
+                ],
+            }
+        )
+    return config
+
+
+def custom_joint_faf_select(ht: hl.Table) -> Dict[str, hl.expr.Expression]:
```
Comment (jkgoodrich) [inline gnomad_qc/v4/create_release/create_release_sites_ht_genomes.py:317]: No change from Mike's so exomes and genomes can use the same function

File: gnomad_qc/v4/create_release/create_release_sites_ht_genomes.py [hunk 3]
```diff
@@ -0,0 +1,790 @@
...
+        - The faf95 field in the grpmax struct is the FAF of the genetic ancestry group with the largest AF (grpmax AF).
+        - The FAF fields within the gen_anc_faf_max struct contains the FAFs from the genetic ancestry group(s) with the largest FAFs
+        - These values aren't necessarily the same; the group with the highest AF for a variant isn't necessarily the group with the highest FAF for a variant
+        - The filtering allele frequencies that are used by the community are the values within the gen_anc_faf_max struct, NOT grpmax FAF, which is why we are dropping grpmax.faf95 and renaming gen_anc_faf_max
+
+    :param ht: Freq Hail Table
+    :return: Select expression dict.
+    """
+    selects = {
+        "grpmax": ht.grpmax.drop("faf95"),
```
Comment (jkgoodrich) [inline gnomad_qc/v4/create_release/create_release_sites_ht_genomes.py:348]: What is the reason for the difference between exomes and genomes here?

File: gnomad_qc/v4/create_release/create_release_sites_ht_genomes.py [hunk 4]
```diff
@@ -0,0 +1,790 @@
...
+        "fafmax": ht.gen_anc_faf_max,
+    }
+
+    return selects
+
+
+def custom_in_silico_select(ht: hl.Table) -> Dict[str, hl.expr.Expression]:
```
Comment (jkgoodrich) [inline gnomad_qc/v4/create_release/create_release_sites_ht_genomes.py:355]: Same as in the exomes PR

File: gnomad_qc/v4/create_release/create_release_sites_ht_genomes.py [hunk 5]
```diff
@@ -0,0 +1,790 @@
...
+    """
+    vep_in_silico = get_sift_polyphen_from_vep(get_vep().ht())
+    selects = {
+        "sift_max": vep_in_silico[ht.key].sift_max,
+        "polyphen_max": vep_in_silico[ht.key].polyphen_max,
+    }
+    return selects
+
+
+def custom_region_flags_select(ht: hl.Table) -> Dict[str, hl.expr.Expression]:
```
Comment (jkgoodrich) [inline gnomad_qc/v4/create_release/create_release_sites_ht_genomes.py:372]: Could also add a data_type parameter here and if "genomes" then add the interval QC. so then exomes and genomes can use the same function

File: gnomad_qc/v4/create_release/create_release_sites_ht_genomes.py [hunk 6]
```diff
@@ -0,0 +1,790 @@
...
+            ht,
+            non_par=True,
+            prob_regions={"lcr": lcr_intervals.ht(), "segdup": seg_dup_intervals.ht()},
+        )
+    }
+
+    return selects
+
+
+def custom_filters_select(ht: hl.Table) -> Dict[str, hl.expr.Expression]:
```
Comment (jkgoodrich) [inline gnomad_qc/v4/create_release/create_release_sites_ht_genomes.py:390]: Same as in the genomes PR

File: gnomad_qc/v4/create_release/create_release_sites_ht_genomes.py [hunk 7]
```diff
@@ -0,0 +1,790 @@
...
+        "mills",
+        "monoallelic",
+        "only_het",
+    ]
+    filters_info_dict = {field: filters[field] for field in filters_info_fields}
+    filters_info_dict.update({**{f"{score_name}": filters[f"{score_name}"]}})
+
+    # Create a dict of the fields from the freq HT that we want to add to the info.
+    freq_ht = get_config().get("freq")["ht"]
+    freq_info_dict = {"inbreeding_coeff": freq_ht[ht.key]["InbreedingCoeff"]}
```
Comment (jkgoodrich) [inline gnomad_qc/v4/create_release/create_release_sites_ht_genomes.py:443]: Probably should have just changed this in the frequency code to be consistent. If we need to remake that table at all, we should change it

File: gnomad_qc/v4/create_release/create_release_sites_ht_genomes.py [hunk 8]
```diff
@@ -0,0 +1,790 @@
...
+    elif filter_name == "IF":
+        name = "isolation_forest_results"
+    else:
+        raise ValueError(f"Filtering method {filter_name} not recognized.")
+    selects = {name: ht.results.annotate(**ht.training_info)}
+
+    return selects
+
+
+def custom_info_select(ht: hl.Table) -> Dict[str, hl.expr.Expression]:
```
Comment (jkgoodrich) [inline gnomad_qc/v4/create_release/create_release_sites_ht_genomes.py:413]: I will look at this function in more detail tomorrow

File: gnomad_qc/v4/create_release/create_release_sites_ht_genomes.py [hunk 9]
```diff
@@ -0,0 +1,790 @@
...
+
+    # Select the info
+    selects = {
+        "info": hl.struct(**info_dict),
+    }
+
+    return selects
+
+
+def custom_vep_select(ht: hl.Table) -> Dict[str, hl.expr.Expression]:
```
Comment (jkgoodrich) [inline gnomad_qc/v4/create_release/create_release_sites_ht_genomes.py:464]: Currently the exact same as genomes

File: gnomad_qc/v4/create_release/create_release_sites_ht_genomes.py [hunk 10]
```diff
@@ -0,0 +1,790 @@
...
+                    "polyphen_prediction",
+                    "polyphen_score",
+                )
+            )
+        )
+    }
+    return selects
+
+
+def get_select_global_fields(ht: hl.Table) -> Dict[str, hl.expr.Expression]:
```
Comment (jkgoodrich) [inline gnomad_qc/v4/create_release/create_release_sites_ht_genomes.py:487]: Currently the exact same as genomes

File: gnomad_qc/v4/create_release/create_release_sites_ht_genomes.py [hunk 11]
```diff
@@ -0,0 +1,790 @@
...
+                global_name = config.get("global_name")
+                select_globals = {global_name: hl.struct(**select_globals)}
+            t_globals.append(select_globals)
+
+    t_globals = reduce(lambda a, b: dict(a, **b), t_globals)
+
+    return t_globals
+
+
+def get_select_fields(
```
Comment (jkgoodrich) [inline gnomad_qc/v4/create_release/create_release_sites_ht_genomes.py:508]: Currently the exact same as genomes

File: gnomad_qc/v4/create_release/create_release_sites_ht_genomes.py [hunk 12]
```diff
@@ -0,0 +1,790 @@
...
+
+    # There is no single path for insilico predictors, so we need to handle this case
+    # separately.
+    if dataset == "in_silico":
+        base_ht = config["ht"]
+    else:
+        ht_path = config["path"]
+        logger.info("Reading in %s", dataset)
+        base_ht = hl.read_table(ht_path, _intervals=_intervals)
+        if dataset == "freq":
```
Comment (jkgoodrich) [inline gnomad_qc/v4/create_release/create_release_sites_ht_genomes.py:558]: Again, if the difference between the datasets are this small, we can just add a data_type parameter and use the function for both

File: gnomad_qc/v4/create_release/create_release_sites_ht_genomes.py [hunk 13]
```diff
@@ -0,0 +1,790 @@
...
+    if "field_name" in config:
+        field_name = config.get("field_name")
+        select_query = {field_name: hl.struct(**select_fields)}
+    else:
+        select_query = select_fields
+
+    return base_ht.select(**select_query)
+
+
+def join_hts(
```
Comment (jkgoodrich) [inline gnomad_qc/v4/create_release/create_release_sites_ht_genomes.py:587]: Currently the exact same as genomes

File: gnomad_qc/v4/create_release/create_release_sites_ht_genomes.py [hunk 14]
```diff
@@ -0,0 +1,790 @@
...
+        args.overwrite,
+    )
+
+    logger.info("Final release HT schema:")
+    ht.describe()
+
+    logger.info("Final variant count: %d", ht.count())
+
+
+if __name__ == "__main__":
```
Comment (jkgoodrich) [inline gnomad_qc/v4/create_release/create_release_sites_ht_genomes.py:738]: add data_type parameter to the other script (and have it renamed without "exomes" in it

**Review by jkgoodrich — CHANGES_REQUESTED** [review]
OK, now that I see how few changes need to be made, I think this should just get added to the script that @mike-w-wilson has put together with a data_type parameter. It can be another PR if his is close to going in, you can just make a branch off of his, but @mike-w-wilson you might want to rename the script with no "exomes" in the name so it can just have a data_type argument added that defaults to exomes.

File: gnomad_qc/v4/create_release/create_release_sites_ht_genomes.py (hunk 3 above)
Comment (KoalaQin) [reply gnomad_qc/v4/create_release/create_release_sites_ht_genomes.py:348]: exomes freq_ht has a nested struct for gnomad & ukbb separately, I think we can use `if` statements.

**Comment by KoalaQin** [comment]
I agree with you, @mike-w-wilson let me know if the exome one is close, after you rename the file, I will make a branch off yours.

**Comment by KoalaQin** [comment]
I have new test run for genomes here: [5de16af8e1114700b96eca66ce838625](https://console.cloud.google.com/dataproc/jobs/5de16af8e1114700b96eca66ce838625?region=us-central1&project=broad-mpg-gnomad) and for exomes here: [b772226de5ce41378c1f743670b3adb8](https://console.cloud.google.com/dataproc/jobs/b772226de5ce41378c1f743670b3adb8?region=us-central1&project=broad-mpg-gnomad)

For some `custom_*_select` functions, I don't need data_type, but I put an optional argument there as a placeholder, maybe you have better solutions for that?

**Comment by KoalaQin** [comment]
Old commits are not important, only the ones I made today matter.

**Review by jkgoodrich — COMMENTED** [review]
Also adding a PR on top of this for some fixes that are needed

**Comment by mike-w-wilson** [comment]
Back to you @jkgoodrich ! I added in a test param to final filters just for testing, noted to drop it as Qin has for combined_faf

**Review by jkgoodrich — APPROVED** [review]
LGTM! Just go through TODO's and remove the ones that were just for testing and then let's merge

**Pull request merged**